### ISSUE TITLE
[CoreRT] Behavior mismatch between CoreCLR and CoreRT Streams in async methods

### ISSUE THREAD

**Issue by jamesqo** [dotnet/corefx#8012]
**TL;DR** This will throw a `TaskCanceledException` in CoreCLR, while it will throw a `OperationCanceledException` in CoreRT:

``` cs
await Stream.Null.FlushAsync(new CancellationToken(true));
```

---

Recently noticed that several async methods in `Stream` and `MemoryStream` have differing implementations when it comes to cancellation tokens in the CoreCLR/CoreFX repos. For example, [here](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/IO/Stream.cs#L894) is the source code of `NullStream.FlushAsync` in CoreCLR, and [here](https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System/IO/Stream.cs#L412) it is in CoreFX. Attempting to await the previous results in a `TaskCanceledException`, and the latter an `OperationCanceledException`.

The fix would be to remove the `ThrowIfCancellationRequested` calls (along with all of the `pragmas`) and replace them with an `IsCancellationRequested` - `Task.FromCanceled` pattern.

@stephentoub, do you know of any reason why the behaviors should differ? If not then I can submit a PR (along with tests) to fix this.

**Comment by stephentoub** [comment]
I don't know why the implementation in corefx doesn't match the one in coreclr; the one in coreclr is more efficient, and it seems like an unnecessary difference for something like this.  I'm not particularly concerned about the difference in exceptions, as TaskCanceledException is an OperationCanceledException (the former derives from the latter), but it's easy enough to make them the same, and doing so seems reasonable unless there's a good reason they were done differently (I'm not aware of any such reasons; @jkotas or @AlfredoMS may be).

**Comment by jkotas** [comment]
The difference was introduced by @ericeil on 6/25/2012 when Stream.cs was forked corefx. The description just says "System.IO.Stream, with unit tests for the *Async methods, which I had to rewrite".

**Comment by jkotas** [comment]
I agree that it would be nice to make them the same.

**Comment by jamesqo** [comment]
@jkotas Great, then; I'll be submitting a PR soon then.

**Comment by stephentoub** [comment]
The implementations have been unified.
